common/src/main/java/io/pravega/common/io/serialization/VersionedSerializer.java

```diff
@@ -0,0 +1,708 @@
+         * to read the serialization header; just the contents.
+         *
+         * @param stream The InputStream to deserialize form.
+         * @return A new instance of TargetType.
+         * @throws IOException If an IO Exception occurred.
+         */
+        TargetType deserializeContents(InputStream stream) throws IOException {
+            ReaderType builder = newBuilder();
+            super.deserializeContents(stream, builder);
+            return builder.build();
+        }
+    }
+
+    //endregion
+
+    //region MultiType
+
+    /**
+     * A VersionedDeserializer that serializes deserializes objects instantiating different types that inherit from a single
+     * base type. This is a meta-serializer in itself, as it composes various other Single-Type serializers into it.
+     *
+     * This should be used in those cases when we have a base (maybe abstract) type and multiple types inheriting from it
+     * that need serialization. A Serializer needs to be implemented for each sub-type and registered into this instance.
+     *
+     * Currently only VersionedSerializer.WithBuilder sub-serializers are supported.
+     *
+     * Example:
+     * <pre>
+     * {@code
+     * class BaseType { ... }
+     *
+     * class SubType1 extends BaseType {
+     *     static class SubType1Builder implements ObjectBuilder<SubType1> { ... }
+     *     static class SubType1Serializer extends VersionedSerializer.WithBuilder<SubType1, SubType1Builder> { ... }
+     * }
+     *
+     * class SubType11 extends SubType1 {
+     *     static class SubType11Builder implements ObjectBuilder<SubType11> { ... }
+     *     static class SubType11Serializer extends VersionedSerializer.WithBuilder<SubType11, SubType11Builder> { ... }
+     * }
+     *
+     * class SubType2 extends BaseType {
+     *     static class SubType2Builder implements ObjectBuilder<SubType2> { ... }
+     *     static class SubType2Serializer extends VersionedSerializer.WithBuilder<SubType2, SubType2Builder> { ... }
+     * }
+     *
+     * class BaseTypeSerializer extends VersionedSerializer.MultiType<BaseType> {
+     *    @Override
+     *    protected void declareSerializers() {
+     *        // Declare sub-serializers here. IDs must be unique, non-changeable (during refactoring) and not necessarily
+     *        // sequential or contiguous.
+     *        serializer(SubType1.class, 0, new SubType1.SubType1Serializer());
+     *        serializer(SubType11.class, 10, new SubType11.SubType11Serializer());
+     *        serializer(SubType2.class, 1, new SubType2.SubType2Serializer());
+     *    }
+     * }
+     * }
+     * </pre>
+     *
+     * @param <BaseType> The base type that all other types will derive from.
+     */
+    public static abstract class MultiType<BaseType> extends VersionedSerializer<BaseType> {
+        private final HashMap<Byte, SerializerInfo> serializersById;
```

Comment:
These aren't thread safe. Maybe they should be immutable maps?

Reply:
I would like to, but if you look at the pattern I chose, they work in the following way:
1. The constructor instantiates them (empty)
2. The constructor invokes the abstract method `declareSerializers`, which is implemented by a subclass
3. The `declareSerializers` subclass will use the `serializer` method to populate these.

Because of 2 and 3, I can't make them `ImmutableMap` and `final` at the same time.

Reply:
Ok make the constructor instantiate two ImmutableMapBuilders and pass them to declareSerailizers() then assign them by calling .build() afterwards. It's all in the constructor, so it can still be final.